octoprint_psucontrol/__init__.py

```diff
@@ -545,16 +549,25 @@ def turn_psu_on(self):
             if self.sensingMethod not in ('GPIO','SYSTEM'):
                 self._noSensing_isPSUOn = True
          
-            time.sleep(0.1 + self.postOnDelay)
+            threading.Timer(0.1 + self.postOnDelay, self.after_psu_on_delay).start()
+
 
+    def after_psu_on_delay(self):
             if self.afterOnGCodeCommands:
                 for command in self._afterOnGCodeCommandsArray:
                     self._printer.commands(command)
 
+            if self.postOnConnect:
+                self._printer.connect()
+
             self.check_psu_state()
+
         
     def turn_psu_off(self):
         if self.switchingMethod == 'GCODE' or self.switchingMethod == 'GPIO' or self.switchingMethod == 'SYSTEM':
+            if self.disconnectOnPowerOff:
+                self._printer.disconnect()

```

Comment:
Because it is always better to do a nice disconnect when the power is still on instead of just switching the PSU off. After PSU off a disconnect is no longer needed as the connection fails anyway.

Reply:
Ok, so this is a case of "how did this ever work as intended"?

Reply:
IMHO… yes. OctoPrint is probably robust enough to handle the disconnect without an issue and does not mind if you try to disconnect an already disconnected printer.
It will start to bring up some issues if another plugin tries to do anything with the printer before disconnecting.